### ISSUE TITLE
populate should set provider to null

### ISSUE THREAD

**Issue by ekryski** [feathersjs/feathers-hooks-common#134]
### Steps to reproduce

1. Setup a hook on a service (ie. messages) that should only be run when `isProvider('external')`.
2. Populate entities from the service (ie. a channel's messages)

I believe it is [this line](https://github.com/feathersjs/feathers-hooks-common/blob/master/src/services/populate.js#L184) that needs to be changed.

### Expected behavior
External only hooks should not run

### Actual behavior
They do run

### System configuration
feathers-hooks-common@3.0.0-pre.1

**Comment by eddyystop** [comment]
The legacy populate hook executed its find's in the same context as itself. If the legacy populate was transported by socketio, then its find's were called as if my socketio. This allowed the hooks for the find's to run properly.

The new populate did not do this originally; it did what you propose. But people started to run into this as a problem and the new populate was changed to act as the old.

Your change would therefore be a breaking change.

Can you do one of:
(a) add a hook before the populate hook that changes hook.params.provider?
(b) clone the hook, modify hook.params.provider can then do `populate(...)(clonedHook)`.

**Comment by ekryski** [comment]
I opted for option "a" thus far.

**Comment by eddyystop** [comment]
I think this issue is resolved, so I'm closing it. Feel free to reopen it.

**Comment by ekryski** [comment]
> But people started to run into this as a problem and the new populate was changed to act as the old.

Do you recall what the problem was that people had? Was it just that populate sensitive data is not getting removed (ie. password when populating a user)? It still seems like a hack to have to put the a hook in front to change the provider.

I don't know what the right solution is here. It **is** an internal call but as @daffl mentioned [here](https://github.com/feathersjs/feathers-hooks-common/pull/136#issuecomment-283159134) it does have implications. Might be worth re-opening but waiting until we can update the `feathers-hooks` and bring it into core.

---

I think the biggest issue with changing it, is if you are populating data from a service and some of that data shouldn't make it to the client we don't have an easy way to ensure the sensitive data is removed. A developer needs to sanitize their data again, which is redundant.

On the flip side having hooks not run because it's not properly being flagged as an internal call might even be worse. Many times we'll be calling services that are not available publicly and have guards preventing external action, but you want to be able to populate from them and certain hooks need to be run for those internal populate calls. In that case it will just straight up error. Lookup tables are a perfect example.

Maybe an updated version of `populate` should always run with `params.provider` **not** being proxied to the other services, but have an `exclude` key in the schema that allows the developer to specify what they want removed during the populate?

**Comment by eddyystop** [comment]
The problem with setting provider `null` on calls `populate` makes is that people have hooks on the called services which depend on the provider. An issue was posted within a day or two of `populate` being released, so a change would be a breaking one.

`populate` sets `hook.params._populate === 'skip'` before calling another service. Perhaps your existing hooks can check for that as a quick fix. **Would that work for you?**

`populate` was not meant to sanitize the data. You yourself came up with the nice idea that populating and sanitizing should be separate, and that's why we have the `sanitize` hook. Perhaps a more systematic approach would expect `populate` to be followed by a `sanitize` hook.

**Comment by ekryski** [comment]
> populate was not meant to sanitize the data. You yourself came up with the nice idea that populating and sanitizing should be separate, and that's why we have the sanitize hook. Perhaps a more systematic approach would expect populate to be followed by a sanitize hook.

Agreed. Which is why I think it **should** be calling other services without `hook.params.provider` being set and if you need to prune sensitive data out of the `hook.result` object then you should do that before you send data out to the client.

You may not even want the service you are populating from to prune data automatically, because you may need to use that data to make other decisions internally on the server side before the response is sent out (I've had this use case), so I think triggering hooks that are meant for external calls is a mistake.

I think the rule should be:

**Any service call originating from the same server should be treated as internal.** If it doesn't originate from the same server side application instance then it is external. So, if you are populating from services on the same app instance then they are internal. If you are populating from a remote instance then it would be an external call. 

> An issue was posted within a day or two of populate being released, so a change would be a breaking one.

I think that's ok. I tend to be more bullish here, but I don't see issues with breaking changes so long as we:

- version properly
- can explain why there is a breaking change
- and give people an adequate migration path

I'm going to re-open this because, although we have a decent temporary solution I think this is something that we should address. Would love to hear thoughts, feelings, impressions from other members of the @feathersjs/core-team.

**Comment by eddyystop** [comment]
I think a better investment of our time right now is to ship the
much-delayed Auk.

**Comment by eddyystop** [comment]
@ekryski I remember reading something you wrote and realizing that it was the likely solution to this issue.

The schema can now be
```javascript
const schema = { include: [
  service: ...
  parentField: ...
  childField: ...
  provider: undefined, // new
]};
```
The provider prop defaults to `hook.params.provider`. Setting it to `undefined` makes the join method call look like its from the server.

Does this work?

**Comment by eddyystop** [comment]
I went ahead with this as its useful in any case (plus wanting to finish with populate for a while). The schema can now be

const schema = { include: [
  service: ...
  parentField: ...
  childField: ...
  provider: undefined, // new
]};
hook.params.provider for the .find call resulting from the join is here set to undefined, hence hooks will process the .find as being made by the server.

By default the provider is set to that of the call to the base record.

Fixed in https://github.com/feathersjs/feathers-hooks-common/pull/172 which will be published as v3.3.0.
